pkg/virt-handler/vm_test.go

```diff
@@ -1656,7 +1657,7 @@ var _ = Describe("VirtualMachineInstance", func() {
 				Expect(arg.(*v1.VirtualMachineInstance).Status.Interfaces[0].Name).To(Equal(podCacheInterface.Iface.Name))
 				Expect(arg.(*v1.VirtualMachineInstance).Status.Interfaces[0].IP).To(Equal(podCacheInterface.PodIP))
 				Expect(arg.(*v1.VirtualMachineInstance).Status.Interfaces[0].IPs[0]).To(Equal(podCacheInterface.PodIP))
```

Comment:
I would change it to check equality to podCacheInterface.PodIPs[0]

Reply:
i would suggest to add another line if so
because IPs[0] should match PodIP, according k8s, as Petr showed, and i added logic for that.

Reply:
That's a unit test, no a func test. It should be equal to whatever you put in podCacheInterface.PodIPs[1] even If [podCacheInterface.PodIP](https://github.com/kubevirt/kubevirt/pull/3958/files#diff-14837489e428608f8ed1f61f9d7242bcR1635) has different value than in the podCacheInterface.PodIPs[1].
 pkg/virt-handler/vm.go doesn't have logic to make sure podCacheInterface.PodIP equals podCacheInterface.PodIPs. Since it is a unit test for pkg/virt-handler/vm.go testing it is wrong.

Reply:
> That's a unit test, not a func test. `arg.(*v1.VirtualMachineInstance).Status.Interfaces[0].IPs[0]` should be equal to whatever you put in `podCacheInterface.PodIPs[0]` even If [podCacheInterface.PodIP](https://github.com/kubevirt/kubevirt/pull/3958/files#diff-14837489e428608f8ed1f61f9d7242bcR1635) has different value than in the `podCacheInterface.PodIPs[0]`.
> pkg/virt-handler/vm.go doesn't have logic to make sure `podCacheInterface.PodIP` equals `podCacheInterface.PodIPs`. Since it is a unit test for pkg/virt-handler/vm.go testing it is wrong.

the PR do have logic for that
https://github.com/kubevirt/kubevirt/pull/3958/files#diff-d3ed8eb8cca020f37cab207c079f2627R124
because K8s API use this logic. it was requested in one of the threads here.

but i wont add if you dont want to, i just think otherwise

Reply:
It's not that I don't want to. It is just wrong. The "logic" is not in vm.go. It is in podInterface.go as you link. Here you [mock](https://github.com/kubevirt/kubevirt/pull/3958/files#diff-14837489e428608f8ed1f61f9d7242bcR1631) the output of podInterface.go. So no "logic" takes place. Meaning, there is no reason to test it.

Reply:
for some reason the link you pasted doesnt jump to the right place in the code,
we do mock the cache creating, but do we also mock the vm.go code ?
anyway i wont add.

Reply:
This code is mocking the `podCacheInterface` file - ```podCacheInterface := network.PodCacheInterface{
				Iface: &v1.Interface{
					Name: interfaceName,
				},
				PodIP: "1.1.1.1",
				PodIPs: [2]string{"1.1.1.1", "fd10:244::8c4c"},
}
podJson, err := json.Marshal(podCacheInterface)
			Expect(err).ToNot(HaveOccurred())
			err = os.MkdirAll(fmt.Sprintf(util.VMIInterfaceDir, vmi.UID), 0755)
			Expect(err).ToNot(HaveOccurred())
			vmiInterfacepath := fmt.Sprintf(util.VMIInterfacepath, vmi.UID, interfaceName)
			f, err := os.Create(vmiInterfacepath)
			Expect(err).ToNot(HaveOccurred())
			_, err = f.WriteString(string(podJson))
			Expect(err).ToNot(HaveOccurred())
			f.Close()```

We don't mock the vm.go code, that's the code we want to test here.

"anyway i wont add" - if you don't agree with my comment, than please give me an argument why this unit test have to verify `Expect(arg.(*v1.VirtualMachineInstance).Status.Interfaces[0].IPs[0]).To(Equal(podCacheInterface.PodIP))`.

Reply:
because i think we do want to test the code of the vm.go, which is the one that read from the file being mocked, and assign in the Status.Interfaces.

I wrote anyway i wont add, because i leave the decision for you, after i tried to explain, sorry that it was implicit that i think we do need to test vm.go code where the actual logic is.

Reply:
But vm.go has no logic that makes sure that `*v1.VirtualMachineInstance).Status.Interfaces[0].IPs[0]` equals to `podCacheInterface.PodIP`. If in the mock you will have the following ` PodIP: "1.1.1.1", PodIPs: [2]string{"2.2.2.2",...`
 `*v1.VirtualMachineInstance).Status.Interfaces[0].IPs[0]` will contain `2.2.2.2` even `podCacheInterface.PodIP` is `1.1.1.1.

Reply:
right, this is why i told e2e, and that it will check the values of the test, 
and can alert people who forget the rule that PodIP == IPs[0]

but it is a semantic of if the test should or not, please advice.
personally i like to add this sort of things that can eliminate problems in the future and not make the unit test too narrow (but of course not too broad)

Reply:
"right, this is why i told e2e" - where you told e2e? What are you talking about?
This is a unit test to a specific method. It should test *only* the functionality of this method.
It shouldn't be used as a reminder to any other functionality not related to this method.

Reply:
"but i think the end result should be tested. (it will even test the mocked values and mock eng)"
ok.